adapters/medianet/medianet.go

```diff
@@ -0,0 +1,118 @@
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+
+	reqJson, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJson,
+		Headers: headers,
+	}}, errs
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponse()
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType, err := getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+// Builder builds a new instance of the Medianet adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	url := buildEndpoint(config.Endpoint, config.ExtraAdapterInfo)
+	return &adapter{
+		endpoint: url,
+	}, nil
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb2.Imp) (openrtb_ext.BidType, error) {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+			return mediaType, nil
+		}
+	}
+
+	return "", &errortypes.BadInput{
+		Message: fmt.Sprintf("Failed to find impression \"%s\" ", impID),
+	}
+}
+
+func buildEndpoint(mnetUrl, hostUrl string) string {
+
+	urlObject, err := url.Parse(mnetUrl)
+	if err != nil || len(hostUrl) == 0 {
+		return mnetUrl
+	}
+	values := urlObject.Query()
+	values.Add("src", hostUrl)
+	urlObject.RawQuery = values.Encode()
+	return urlObject.String()
```

Comment:
Could we consider additional scenarios inside `adapters/medianet/medianet_test.go` in order to test this function further? 

Current coverage:
![image](https://user-images.githubusercontent.com/15863840/151048618-944f0814-dfef-47bf-af06-f11f688fcd71.png)

Reply:
sure let me add those.

Reply:
Done, Thanks @guscarreon

Reply:
Thanks for adding coverage @eshaanmangal. Another ask in regards of this function: If `hostUrl` comes empty there seems to not be need to call `url.Parse(mnetUrl)` at all. Can we slightly optimize?
```
108   func buildEndpoint(mnetUrl, hostUrl string) string {
109  
    +     if len(hostUrl) == 0 {
    +         return mnetUrl
    +     }
110       urlObject, err := url.Parse(mnetUrl)
111 -     if err != nil || len(hostUrl) == 0 {
    +     if err != nil {
112           return mnetUrl
113       }
114       values := urlObject.Query()
115       values.Add("src", hostUrl)
116       urlObject.RawQuery = values.Encode()
117       return urlObject.String()
118   }
adapters/medianet/medianet.go
```

Reply:
yes that's valid point but the reason I changed it assuming (might be over-thinking), what if some custom implementation of prebid server (manage prebid) set spammed medianet url in pbs.yaml (config) file